src/utils/validators.js

```diff
@@ -31,50 +30,50 @@ export const validateConditionalRules = (key, attributeTypes, attributes) => {
  * @return {Validator}]
  */
 const createValidateErrors = (type) => (obj, rules) => {
-  const errors = [];
-  Object.keys(rules).forEach((key) => {
-    if (Object.prototype.hasOwnProperty.call(rules, key)) {
-      const rule = rules[key];
-      const isRequired = rules[key].required;
-      const isRequiredInType = includes(rule.requiredIn, type);
+  const keys = Object.keys(rules).filter((key) => Object.hasOwn(rules, key))
+
+  const errors = keys.filter((key) => {
+    const rule = rules[key];
+
+    const isRequired = rules[key].required;
+    const isRequiredInType = includes(rule.requiredIn, type);
+    const objHasRuleAttribute = Boolean(obj.attributes[key]);
+
+    const isAttributeAllowedInType = includes(rules[key].allowedIn, type);
+    const isValid =
+      includes(
+        rule.values.map(({ value }) => value),
+        obj.attributes[key]
+      ) || rule.values.length === 0;
+    const allowValuesOutsideChoices = includes(
+      rule.allowValuesOutsideChoicesIn,
+      type
+    );
+
+    return (isRequired && isRequiredInType && !objHasRuleAttribute) ||
+      (objHasRuleAttribute && !isValid && !allowValuesOutsideChoices) ||
+      (!isAttributeAllowedInType && objHasRuleAttribute)
+  })
+
+  const conditionallyRequired = keys
+    .filter(((key) => rules[key].requiredIf.length > 0))
```

Comment:
I believe you can leave out one set of brackets here.

Reply:
Whoops, will fix 👀

Reply:
https://github.com/City-of-Helsinki/helerm-ui/pull/493/commits/fbb6b21dc6c62a650329de64a01bb83219537b41